lib/discordgo/gateway.go

```diff
@@ -797,52 +797,93 @@ func (g *GatewayConnection) Close() error {
 	return nil
 }
 
-func newUpdateStatusData(idle int, gameType GameType, game, url string) *UpdateStatusData {
+func newUpdateStatusData(activityType ActivityType, statusType Status, statusText, streamingUrl string) *UpdateStatusData {
 	usd := &UpdateStatusData{
-		Status: "online",
+		Status: statusType,
 	}
 
-	if idle > 0 {
+	/*if idle > 0 {
 		usd.IdleSince = &idle
-	}
-
-	if game != "" {
-		usd.Game = &Game{
-			Name: game,
-			Type: gameType,
-			URL:  url,
+	}*/
+
+	if statusText != "" {
+		usd.Activity = &Activity{
+			Name:  statusText,
+			State: statusText,
+			Type:  activityType,
+			URL:   streamingUrl,
 		}
 	}
-
 	return usd
 }
 
 // UpdateStatus is used to update the user's status.
-// If idle>0 then set status to idle.
-// If game!="" then set game.
-// if otherwise, set status to active, and no game.
-func (s *Session) UpdateStatus(idle int, game string) (err error) {
-	return s.UpdateStatusComplex(*newUpdateStatusData(idle, GameTypeGame, game, ""))
+// Set the custom status to statusText.
+// Set the online status to statusType.
+func (s *Session) UpdateStatus(activityType string, statusType Status, statusText, streamingUrl string) (err error) {
```

Comment:
The presence of `streamingUrl` is a sign of bad API design IMO -- if I want to set playing status, I still need to pass an empty string. So I'm not convinced this function is appropriate. WDYT?

Reply:
...that said perhaps this is a necessary evil (and it's not too bad anyway) if we want to avoid code duplication. That said, if we're keeping this function I'd suggest splitting `string -> ActivityType` into one function `ParseActivityType` and then change `UpdateStatus` to accept an `ActivityType` instead of a `string`.

Reply:
> The presence of `streamingUrl` is a sign of bad API design IMO -- if I want to set playing status, I still need to pass an empty string. So I'm not convinced this function signature is appropriate. WDYT?

In this case, since it is the top-level function that's supposed to handle deciding for the user, IMO this is the simplest way

> ...that said perhaps this is a necessary evil (and it's not too bad anyway) if we want to avoid code duplication.
yeah this

> That said, if we're keeping this function I'd suggest splitting `string -> ActivityType` into one function `ParseActivityType` and then change `UpdateStatus` to accept an `ActivityType` instead of a `string`.
will do :+1:

Reply:
Wait why? Right now it's not being converted because it decides the function right away, wouldn't it just be extra steps to convert to ActivityType first? Or are you recommending to remove `s.Update{Type}Status` entirely, and have `UpdateStatus` make the call to `UpdateStatusComplex` right away?

Reply:
Sure, either would work.

Reply:
Alright, I separated out that conversion into another function, did leave the `s.Update{Type}Status` in place as well though.